audit/base/base.go

```diff
@@ -510,23 +554,42 @@ func (c *Client) write(ctx context.Context, msg msgs.Msg) {
 		c.setErr(fmt.Errorf("%w: %w", err, ErrConnection))
 		return
 	}
+	c.metrics.msgsSent.Add(ctx, 1)
 	c.successSend = true
 }
 
+// msgRequeueOrDrop requeues the message if there is room in the queue. If there is no room, the message is dropped.
+func (c *Client) msgRequeueOrDrop(ctx context.Context, msg msgs.Msg, err error) {
+	select {
+	case c.sendCh <- SendMsg{Ctx: ctx, Msg: msg}:
+		c.metrics.requeuedCounter.Add(1)
+		c.metrics.msgsRequeued.Add(ctx, 1)
+	default:
+		c.metrics.msgsDropped.Add(ctx, 1)
+		c.log.Error(fmt.Sprintf("audit message dropped due to queue being full: %v", err))
+	}
+}
+
 // setErr sets the error for the client if the error is not already set. If the error being set it nil,
 // it will be set to nil.
 func (c *Client) setErr(err error) {
 	c.setErrMu.Lock()
 	defer c.setErrMu.Unlock()
 
+	// This resets the error if the error passed is nil.
 	if err == nil {
 		c.err.Store(nil)
+		return
 	}
 
+	// If there already is an error, we don't need to overwrite it.
 	if c.err.Load() != nil {
 		return
 	}
 
+	// This sets the conn to nil. This prevents any further writes to the connection until it gets replaced.
+	c.conn.Store(nil)
```

Comment:
Does this produce a conn-leak?

Before, client was preserved and then closed in `Reset` which was (theoretically) eventually called by the outer client. Now, when `Reset` is called if there was an error on this client the conn will be nil and there won't be anything to close, but the actual ref was never closed?

Do we need to close it here?

Reply:
Yeah, looks like we'd leak.  I've updated the code to close it here.